File: process_cliopatria.js
#!/usr/bin/env node
/**
 * Process Cliopatria GeoJSON → browser-optimized data.
 * Two-stage strategy:
 *   1. Area >= 200K km²  → keep (major polities)
 *   2. Duration >= 50yr  → keep if Area >= 20K km² (long-lived entities)
 * Then simplify + round coords to compress.
 */
const fs = require('fs');
const INPUT = 'cliopatria_polities_only.geojson';

const COLORS = {
  europe: "#3b82f6", middle_east: "#f59e0b", east_asia: "#ef4444",
  south_asia: "#22c55e", southeast_asia: "#06b6d4", africa: "#a855f7",
  americas: "#ec4899", central_asia: "#f97316", steppe: "#78716c",
  default: "#6b7280",
};

const REGION_KEYWORDS = {
  east_asia: ["china","chinese","han dynast","tang dynast","song dynast","yuan dynast",
    "ming dynast","qing dynast","jin dynast","sui dynast","xia dynast","shang dynast",
    "zhou dynast","qin dynast","korea","joseon","goryeo","silla","baekje","goguryeo",
    "japan","yamato","nara","heian","tokugawa","edo","meiji","taiwan","ryukyu",
    "tibet","uyghur","tangut","khitan","jurchen","manchu","mongol empire","great mongol",
    "buyeo","nanzhao","dali","chu","yan","zhao","wuwei","tufan"],
  south_asia: ["india","maurya","gupta","mughal","delhi sultan","chola","pallava",
    "vijayanagara","maratha","persia","iran","safavid","qajar","parthia","sasanid",
    "sasanian","timurid","kushan","gandhara","gurjara","pratihara","rastrakuta",
    "chalukya","hoysala","kakatiya","kalinga","magadha","bengal","deccan","sindh",
    "kabul","merv","balkh","transoxiana","sogdiana","bactria"],
  europe: ["roman republ","roman empi","byzantin","byzantium","frank","carolingian",
    "holy roman","spain","portugal","france","england","britain","british","scotland",
    "ireland","germany","prussia","austria","hungary","poland","lithuania","russia",
    "sweden","norway","denmark","dutch","netherlands","belgium","italy","venice",
    "genoa","papal","ottoman","serbia","bulgaria","croatia","bohemia","romania",
    "ukraine","finland","estonia","latvia","albania","greece","gothic","vandal",
    "visigoth","ostrogoth","lombard","saxon","norman","burgundy","savoy","aragon",
    "castile","moorish","teutonic","crusader","hispania","gaul","dacia","pannonia"],
  middle_east: ["arab","abbasid","umayyad","fatimid","ayyubid","mamluk",
    "crusader","assyria","mesopotamia","levant","syria","israel","hejaz","saudi",
    "yemen","iraq","turkish","seljuq","rashidun","phoenicia","carthage","akkad",
    "sumer","elam","urartu","palmyra","timurid","babylon"],
  africa: ["egypt","kush","aksum","ethiopia","mali","songhai","ghana","zulu",
    "ashanti","kongo","swahili","morocco","tunisia","algeria","libya","sudan",
    "nubia","madagascar","carthage","numidia","mauretania","axum","kanem","bornu",
    "luba","kuba","lunda","chokwe","mutapa","matamba","kambongo","loango",
    "kandy","ceylon","singapore"],
  americas: ["aztec","maya","inca","muisca","mississippian","united state",
    "american","canada","brazil","mexico","argentina","chile","colombia","peru",
    "venezuela","bolivia","toltec","zapotec","mixtec","tarascan"],
  southeast_asia: ["siam","thai","burma","burmese","khmer","angkor","champa",
    "dai viet","vietnam","malacca","malay","srivijaya","majapahit","mataram",
    "lanna","ayutthaya","sukhothai","lao","shan","mon","arakan"],
  central_asia: ["turkic","uighur","karakhanid","ghaznavid","khwarazm","sogdiana",
    "ferghana","transoxiana","kazakh","uzbek","turkmen","kyrgyz","tajik","bukhara",
    "samarkand","khiva","kokand","dzungar","karakhitay","qara khitai","khazar"],
  steppe: ["hunnic","xiongnu","scythian","sarmatian","hun","huna","hephthalite",
    "gokturk","turgesh","mongol","tatar","khereid","naiman","kerait","kipchak",
    "golden horde","white horde","chagatai","pecheneg","cuman","bulgar","avar",
    "alan","gothic","oyirad","oirat","torghut","khoid","choros","khoshut"],
};

function assignColor(name) {
  const lower = name.toLowerCase();
  for (const [region, keywords] of Object.entries(REGION_KEYWORDS)) {
    for (const kw of keywords) {
      if (lower.includes(kw)) return COLORS[region] || COLORS.default;
    }
  }
  return COLORS.default;
}

function roundCoords(coords) {
  return coords.map(c => [Math.round(c[0] * 10) / 10, Math.round(c[1] * 10) / 10]);
}

function simplify(coords, every) {
  if (coords.length <= 4) return coords;
  const r = [];
  for (let i = 0; i < coords.length; i += every) r.push(coords[i]);
  const last = coords[coords.length - 1];
  if (r[r.length - 1][0] !== last[0] || r[r.length - 1][1] !== last[1]) r.push(last);
  return r;
}

function process() {
  console.log('Loading GeoJSON...');
  const data = JSON.parse(fs.readFileSync(INPUT, 'utf-8'));
  console.log(`Total features: ${data.features.length}`);

  const features = [];
  let skipped = 0;
  for (const feat of data.features) {
    const p = feat.properties || {};
    const g = feat.geometry || {};
    if (p.Type === 'RELATION') continue;

    const area = p.Area || 0;
    const duration = (p.ToYear || 0) - (p.FromYear || 0);
    // Keep if large OR long-lived
    if (area < 200000 && duration < 50) { skipped++; continue; }

    const color = assignColor(p.Name || '');
    const entry = { n: p.Name, fy: p.FromYear, ty: p.ToYear, c: color };

    const simplifyFactor = area < 500000 ? 40 : 25;
    if (g.type === 'Polygon') {
      entry.g = (g.coordinates || []).map(r => roundCoords(simplify(r, simplifyFactor)));
    } else if (g.type === 'MultiPolygon') {
      entry.gm = (g.coordinates || []).map(poly => poly.map(r => roundCoords(simplify(r, simplifyFactor))));
    } else continue;

    features.push(entry);
  }

  console.log(`Features kept: ${features.length} (skipped ${skipped})`);

  const MIN_YEAR = -5000, MAX_YEAR = 2050;
  const yearIndex = new Array(MAX_YEAR - MIN_YEAR + 1).fill(null).map(() => []);
  for (let i = 0; i < features.length; i++) {
    const f = features[i];
    const start = Math.max(f.fy, MIN_YEAR);
    const end = Math.min(f.ty, MAX_YEAR);
    for (let y = start; y <= end; y++) {
      yearIndex[y - MIN_YEAR].push(i);
    }
  }

  fs.writeFileSync('cliopatria_features.js',
    '// Cliopatria (Seshat) — CC BY 4.0\nvar CLIOPATRIA_FEATURES=' + JSON.stringify(features) + ';\n', 'utf-8');

  const indexData = [];
  for (let i = 0; i < yearIndex.length; i++) {
    if (yearIndex[i].length > 0) {
      indexData.push([i + MIN_YEAR, yearIndex[i]]);
    }
  }
  fs.writeFileSync('cliopatria_index.js',
    'var CLIOPATRIA_INDEX=' + JSON.stringify(indexData) + ';\n', 'utf-8');

  const fSize = fs.statSync('cliopatria_features.js').size;
  const iSize = fs.statSync('cliopatria_index.js').size;
  console.log(`Features: ${(fSize / 1024 / 1024).toFixed(1)} MB`);
  console.log(`Index: ${(iSize / 1024).toFixed(0)} KB`);
  console.log(`Total: ${((fSize + iSize) / 1024 / 1024).toFixed(1)} MB`);
}

process();

File: process_cliopatria.py
#!/usr/bin/env python3
"""
Simplify Cliopatria GeoJSON for browser use.
- Keeps only major polities (Area > 50000 km² or well-known names)
- Simplifies polygon coordinates
- Adds color assignments
- Outputs a compact JSON script file
"""
import json, sys, math

INPUT = "cliopatria_polities_only.geojson"
OUTPUT = "cliopatria_data.js"

# Color palette for different regions/eras
COLORS = {
    "default": "#888888",
    "europe": "#3b82f6",
    "middle_east": "#f59e0b",
    "east_asia": "#ef4444",
    "south_asia": "#22c55e",
    "southeast_asia": "#06b6d4",
    "africa": "#a855f7",
    "americas": "#ec4899",
    "central_asia": "#f97316",
    "steppe": "#78716c",
    "oceania": "#14b8a6",
}

# Keywords to map polity names to regions
REGION_KEYWORDS = {
    "east_asia": ["china", "chinese", "han dynasty", "tang dynasty", "song dynasty", "yuan dynasty",
                  "ming dynasty", "qing dynasty", "jin dynasty", "wei", "shu", "wu",
                  "qi", "liang", "chen", "sui dynasty", "xia dynasty", "shang dynasty",
                  "zhou dynasty", "qin dynasty", "korea", "joseon", "goryeo", "silla",
                  "baekje", "goguryeo", "japan", "yamato", "nara", "heian", "tokugawa",
                  "edo", "meiji", "taiwan", "ryukyu", "ainu", "tibet", "uyghur",
                  "tangut", "khitan", "jurchen", "manchuria", "manchu", "mongol",
                  "mongolia", "yuan", "mongol empire", "great mongolia", "xinjiang",
                  "dzungar"],
    "south_asia": ["india", "maurya", "gupta", "mughal", "delhi sultanate", "chola",
                   "pallava", "vijayanagara", "maratha", "sikh", "afghanistan",
                   "persia", "iran", "safavid", "qajar", "pahlavi", "parthia",
                   "sasanid", "sasanian", "seljuk", "khwarezm", "timurid"],
    "europe": ["rome", "roman", "byzantine", "byzantium", "frank", "carolingian",
               "holy roman", "byzantine", "spain", "portugal", "france", "england",
               "britain", "british", "scotland", "ireland", "germany", "prussia",
               "austria", "hungary", "poland", "lithuania", "russia", "russian",
               "sweden", "norway", "denmark", "dutch", "netherlands", "belgium",
               "italy", "venice", "genoa", "papal", "OTTOMAN", "ottoman",
               "serbia", "bulgaria", "croatia", "bohemia", "czech", "slovak",
               "romania", "moldova", "ukraine", "finland", "estonia", "latvia",
               "lithuania", "albania", "montenegro", "bosnia", "macedonia",
               "greece", "athens", "sparta", "thebes", "argos", "corinth",
               "wallachia", "transylvania", "crimea", "gothic", "vandal",
               "visigoth", "ostrogoth", "lombard", "saxon", "viking", "norman",
               "burgundy", "savoy", "swiss", "catalonia", "aragon", "castile",
               "leon", "navarre", "granada", "moorish", "alusanne", "bavaria",
               "saxony", "brandenburg", "hanseatic", "teutonic", "hospitaller",
               "venetian", "genoese", "crusader", "ionian", "crete"],
    "middle_east": ["arab", "abbasid", "umayyad", "fatimid", "ayyubid",
                    "mamluk", "seljuk", "crusader", "babylon", "assyria",
                    "mesopotamia", "levant", "syria", "jordan", "israel",
                    "hejaz", "nedjd", "saudi", "yemen", "oman", "bahrain",
                    "qatar", "kwait", "iraq", "turkey", "turkish", "seljuq",
                    "danishmend", "artuqid", "zengid", "rum", "hafsid",
                    "saadi", "alaouite", "cherifian", "rashidun"],
    "africa": ["egypt", "kush", "aksum", "ethiopia", "ethiopian", "mali",
               "songhai", "ghana", "benin", "zulu", "ashanti", "kongo",
               "swahili", "zanzibar", "morocco", "tunisia", "algeria",
               "libya", "sudan", "nubia", "somali", "madagascar", "african",
               "carthage", "numidia", "mauretania", "axum", " columnHeader=",
               "kanem", "bornu", "hausa", "nupe", "oyo", "ifẹ", "ibadan",
               "loango", "lunda", "chokwe", "mutapa", "rozwi", "ndebele",
               "matebele", "shona", "makololo", "tswana", "sotho", "tsonga",
               "venda", "swazi", "xhosa", "thembu", "pondo", "griqua",
               "khoikhoi", "san"],
    "americas": ["aztec", "maya", "inca", "muisca", "mound", "mississippian",
                 "hohokam", "ancestral", "pueblo", "aztec", "toltec", "olmec",
                 "zapotec", "mixtec", "tarascan", "totonac", "huastec",
                 "united states", "american", "canada", "brazil", "mexico",
                 "argentina", "chile", "colombia", "peru", "venezuela",
                 "bolivia", "paraguay", "uruguay", "ecuador", "guyana",
                 "suriname", "french guiana", "cuba", "jamaica", "haiti",
                 "dominican", "trinidad", "puerto rico", "belize", "guatemala",
                 "honduras", "el salvador", "nicaragua", "costa rica", "panama"],
    "southeast_asia": ["siam", "thai", "burma", "burmese", "myanmar", "khmer",
                       "angkor", "champa", "dai viet", "vietnam", "viet",
                       "malacca", "malay", "srivijaya", "majapahit", "mataram",
                       "ceylon", "singapore", "borneo", "sulu", "maguindanao",
                       "tondo", "cebu", "palawan", "batanes", "luzon",
                       "visayas", "mindanao", "sulu", "brunei", "sarawak",
                       "sabah", "kalimantan", "sumatra", "java", "bali",
                       "sulawesi", "maluku", "timor", "papua", "moluccas",
                       "aceh", "minangkabau", "batak", "bugis", "makassar",
                       "madura", "javanese", "balinese", "khmer", "lao",
                       "shan", "mon", "arakan", "manipur", "kachin", "chin",
                       "karen", "kayin", "shan", "wa", "palaung", "rumai"],
    "central_asia": ["turkic", "uighur", "karakhanid", "qarakhanid", "ghaznavid",
                     "ghurid", "khwarazm", "chorasmia", "sogdiana", "bactria",
                     "ferghana", "transoxiana", "kazakh", "uzbek", "turkmen",
                     "kyrgyz", "tajik", "bukhara", "samarkand", "khiva",
                     "kokand", "dzungar", "kazakh", "kirghiz", "burut"],
    "steppe": ["hunnic", "hunnic", "xiongnu", "silk road", "scythian",
               "sarmatian", "alan", "gothic", "avar", "bulgar", "turkic",
               "khazar", "pecheneg", "cuman", "kipchak", "nogai", "crimean",
               "golden horde", "white horde", "blue horde", "chagatai",
               "ogir", "naiman", "merkit", "kerait", "onggirat"],
}

def assign_color(name, area):
    name_lower = name.lower()
    for region, keywords in REGION_KEYWORDS.items():
        for kw in keywords:
            if kw.lower() in name_lower:
                return COLORS.get(region, COLORS["default"])
    if area and area > 2000000:
        return COLORS["europe"]
    return COLORS["default"]

def simplify_coords(coords, keep_every=3):
    """Keep every Nth coordinate to reduce size."""
    if len(coords) <= 6:
        return coords
    return [coords[i] for i in range(0, len(coords), keep_every)]

def process():
    print("Loading GeoJSON...")
    with open(INPUT, "r", encoding="utf-8") as f:
        data = json.load(f)

    features = data["features"]
    print(f"Total features: {len(features)}")

    # Filter and simplify
    result = []
    for feat in features:
        props = feat.get("properties", {})
        geom = feat.get("geometry", {})
        name = props.get("Name", "Unknown")
        from_year = props.get("FromYear", 0)
        to_year = props.get("ToYear", 0)
        area = props.get("Area", 0)
        geo_type = props.get("Type", "POLITY")

        # Skip RELATION types (subdivisions, etc.)
        if geo_type == "RELATION":
            continue

        # Keep polities with significant area or notable names
        if area < 30000:
            continue

        color = assign_color(name, area)

        # Process geometry
        geom_type = geom.get("type", "")
        if geom_type == "Polygon":
            rings = geom.get("coordinates", [])
            simplified = []
            for ring in rings:
                simplified.append(simplify_coords(ring, keep_every=2))
            result.append({
                "n": name,
                "fy": from_year,
                "ty": to_year,
                "a": round(area),
                "c": color,
                "g": simplified
            })
        elif geom_type == "MultiPolygon":
            polygons = geom.get("coordinates", [])
            simplified = []
            for poly in polygons:
                rings = []
                for ring in poly:
                    rings.append(simplify_coords(ring, keep_every=2))
                simplified.append(rings)
            result.append({
                "n": name,
                "fy": from_year,
                "ty": to_year,
                "a": round(area),
                "c": color,
                "gm": simplified
            })

    print(f"Filtered features: {len(result)}")

    # Write as JS module
    with open(OUTPUT, "w", encoding="utf-8") as f:
        f.write("// Auto-generated from Cliopatria dataset (simplified)\n")
        f.write("// Source: Seshat Global History Databank\n")
        f.write("// License: CC BY 4.0\n")
        f.write("const CLIOPATRIA_DATA = ")
        json.dump(result, f, ensure_ascii=False, separators=(",", ":"))
        f.write(";\n")

    file_size = len(open(OUTPUT, "r", encoding="utf-8").read())
    print(f"Output: {OUTPUT} ({file_size / 1024 / 1024:.1f} MB)")

if __name__ == "__main__":
    process()
